lib/spack/spack/spec.py

```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
                         new_dependencies.add(edge)
             spec._dependencies = new_dependencies
 
+    def _virtuals_provided(self, root):
+        """Return set of virtuals provided by self in the context of root"""
+        if root is self:
+            # Could be using any virtual the package can provide
+            return set(self.package.virtuals_provided)
+
+        virtuals = set()
+        hashes = [s.dag_hash() for s in root.traverse()]
+        in_edges = set(
+            [edge for edge in self.edges_from_dependents() if edge.parent.dag_hash() in hashes]
+        )
+        for edge in in_edges:
+            virtuals |= set(edge.virtuals)
+
+        return virtuals
+
+    def _splice_match(self, other, self_root, other_root):
+        """Return True if other is a match for self in a splice of other_root into self_root"""
+        if other.name == self.name:
+            return True
+
+        return bool(
+            self._virtuals_provided(self_root)
+            and self._virtuals_provided(self_root) <= other._virtuals_provided(other_root)
+        )
+
+    def _splice_detach_and_add_dependents(self, replacement, context):
+        """Helper method for Spec._splice_helper.
+
+        replacement is a node to splice in, context is the scope of dependents to consider relevant
+        to this splice."""
+        # Update build_spec attributes for all transitive dependents
+        # before we start changing their dependencies
+        ancestors_in_context = [
+            a
+            for a in self.traverse(root=False, direction="parents")
+            if a in context.traverse(deptype=dt.LINK | dt.RUN)
+        ]
+        for ancestor in ancestors_in_context:
+            # Only set it if it hasn't been spliced before
+            ancestor._build_spec = ancestor._build_spec or ancestor.copy()
+            ancestor.clear_cached_hashes(ignore=(ht.package_hash.attr,))
+
+        # For each direct dependent in the link/run graph, replace the dependency on
+        # node with one on replacement
+        # For each build dependent, restrict the edge to build-only
+        for edge in self.edges_from_dependents():
+            if edge.parent not in ancestors_in_context:
+                continue
+            build_dep = edge.depflag & dt.BUILD
+            other_dep = edge.depflag & ~dt.BUILD
+            if build_dep:
+                parent_edge = [e for e in edge.parent._dependencies[self.name] if e.spec is self]
+                assert len(parent_edge) == 1
+
+                edge.depflag = dt.BUILD
+                parent_edge[0].depflag = dt.BUILD
+            else:
+                edge.parent._dependencies.edges[self.name].remove(edge)
+                self._dependents.edges[edge.parent.name].remove(edge)
+
+            if other_dep:
+                edge.parent._add_dependency(replacement, depflag=other_dep, virtuals=edge.virtuals)
+
+    def get_analogues(self, candidates_generator, self_root, other_root):
```

Comment:
Because I'd never seen the US spelling in the context of being analogous, only in contrast to being digital, and didn't realize it was an option. I'll update it for US spelling.